---
title: "Proof of concept Zstandard user-provided dictionary compression for FASTA files"
format:
  plos-pdf:
    number-sections: false
    journal:
      # This is an identifier for the target journal: 
      # from https://plos.org/resources/writing-center/ following submission guidelines link, the identifier is the part of the URL after https://journals.plos.org/<id>/s/submission-guidelines
      id: plosbiology
    include-in-header: |
      % Remove comment for double spacing
      % \usepackage{setspace} 
      % \doublespacing
author:
  - name: Michael Persico
    equal-contributor: false
    affiliations:
      - ref: aff1
    notes: |
      "Current Address: Dept/Program/Center, Institution Name, City, State, Country"
    corresponding: false
    email: correspondingauthor@institute.edu
author-notes:
  equal-contributor: These authors contributed equally to this work.
  deceased: Deceased
  group: Membership list can be found in the Acknowledgments sections
affiliations:
  - id: aff1
    name: Department of Biology, Concordia University
    city: Montreal
    state: Quebec
    country: Canada
abstract: |
  ## Background

  Zstandard (Zstd) represents a universal, lossless data compression mechanism that is highly configurable and is aimed at coupling high compression ratios with fast compression/decompression performance. Previous studies have paired specific Zstd configurations with various file formats in bioinformatics to reduce total data volume. This paper presents a "training mode" pipeline, written in the Julia programming language, wherein a custom compression dictionary is generated from a sample FASTA set in order to explore further compression improvements and compare them to the compression performance of Xz, Zlib, Bzip2, and Lz4 universal compressors.
  
  ## Results

  ## Conclusions
bibliography: bibliography.bib 
---

# Introduction

The explosion of biological data has represented a significant topic of research, with a number of challenges presented over subsequent generations of technological development in regards to the management of  the increasing volume and complexity of data[@d2018high;@li2014big]. In response, emerging trends in data management have lead to the development of novel, scalable methods for the efficient transmission and storage of large amounts of data[@sais2022intelligent]. With a potentially exponential quantity of files, datasets, and other data resources to be handled, data compression represents a method for reducing overall resource size by encoding the original data into a compact form, thus helping to ease storage requirements [@jayasankar2021survey]. Research into data compression in the context of biological data began to pick up near the turn of the 21st century as universal compression algorithms at the time were not considered ideal for compressing DNA or RNA sequence data well, which led to the introduction of purpose-built algorithms that addressed the unique attributes of genomics data[@grumbach1994new]. At the same time, new file formats were introduced, either text-based or binary-based, for more accurate structuring and representation of biological data, complementing new software tools[@lipman1985rapid;@mills2014common].

The FASTA file format is a legacy of the original FASTA program from 1985 for finding sequence similarities with a query sequence[@lipman1985rapid]. Each file can possess multiple sequences, each paired with a description line distinguished with a ">" symbol followed by arbitary text, usually a name and summary description, on the same line. It is a commonly supported file format in bioinformatics and has been the target for optimized data compressors in competition with one another. The DELIMINATE lossless algorithm was first proposed in 2012, wherein header and sequence data are separated into DELIM-1 and DELIM-2 variants and a two-phase process is pursued involving delta encoding, progressive elimination of nucleotide characters, and 7-Zip archiving[@mohammed2012deliminate]. The claims of better compression/decompression performance of FASTA files were soon rivaled by the introduction of the MFCompress tool, again separating headers and sequence data but instead relying on probabilistic models to encode the data[@pinho2014mfcompress], which was countered by the Nucleotide Archival Format, a novel file format which is noteworthy in this context for the inclusion a Zstandard compression step[@kryukov2019nucleotide].


# Materials and methods 

A list of all direct and indirect dependencies can be found in the repository's Manifest.toml file.

# Results

# Discussion

# Conclusion

# Supporting information 

::: {.supp}
## S1 Fig. {#s1-fig}

Bold the title sentence.

Add descriptive text after the title of the item (optional).
:::

::: {.supp}
## S2 Fig. {#s2-fig}

Lorem ipsum.

Maecenas convallis mauris sit amet sem ultrices gravida. Etiam eget sapien nibh. Sed ac ipsum eget enim egestas ullamcorper nec euismod ligula. Curabitur fringilla pulvinar lectus consectetur pellentesque.
:::


::: {.supp}
## S1 File. {#s1-file}

Lorem ipsum.

:::

::: {.supp}

## S1 Video. {#s1-video}

Lorem ipsum.

Maecenas convallis mauris sit amet sem ultrices gravida. Etiam eget sapien nibh. Sed ac ipsum eget enim egestas ullamcorper nec euismod ligula. Curabitur fringilla pulvinar lectus consectetur pellentesque.
:::

::: {.supp}
## S1 Appendix. {#s1-appendix}

Lorem ipsum.

Maecenas convallis mauris sit amet sem ultrices gravida. Etiam eget sapien nibh. Sed ac ipsum eget enim egestas ullamcorper nec euismod ligula. Curabitur fringilla pulvinar lectus consectetur pellentesque.
:::

::: {.supp}
## S1 Table. {#s1-table}

Lorem ipsum.

Maecenas convallis mauris sit amet sem ultrices gravida. Etiam eget sapien nibh. Sed ac ipsum eget enim egestas ullamcorper nec euismod ligula. Curabitur fringilla pulvinar lectus consectetur pellentesque.
:::

# Acknowledgments

I would like to express my sincere gratitude to Professor David Walsh for his advice that helped shape the research as it progressed; My friends and family that supported me throughout my studies; The Julia community for their support before and throughout the writing of this paper; The Quarto developers for producing the Quarto publishing system[@Allaire_Quarto_2022] along with the PLOS template used for this paper; Luca Di Maio and contributors to the Distrobox tool for the ease of setting up the containerized development environments[@maio]; Maximiliano Sandoval and contributors to the Citations app for managing the paper's bibliography[@sandoval], and all other persons that had indirectly assisted through their programs and research.
